### PR TITLE
RavenDB-15715 - Data compression for TCP connections

### PR DESCRIPTION
**Pull request by shiranshalom**

### CODE REVIEW

File: src/Raven.Client/Documents/Operations/Attachments/GetAttachmentsOperation.cs [hunk 1]
```diff
@@ -13,6 +13,7 @@
 using Sparrow;
 using Sparrow.Json;
 using Sparrow.Json.Parsing;
+using Sparrow.Utils;
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Operations/Attachments/GetAttachmentsOperation.cs:16]: remove?

File: src/Raven.Client/Raven.Client.csproj [hunk 2]
```diff
@@ -3,7 +3,7 @@
     <Version></Version>
     <Description>RavenDB Client is the client library for accessing RavenDB</Description>
     <Authors>Hibernating Rhinos</Authors>
-    <TargetFrameworks>net5.0;netcoreapp3.1;netcoreapp2.1;netstandard2.1;netstandard2.0</TargetFrameworks>
```
Comment (ppekrol) [inline src/Raven.Client/Raven.Client.csproj]: revert

File: src/Raven.Embedded/EmbeddedServer.cs [hunk 3]
```diff
@@ -6,12 +6,8 @@
 using System.Threading.Tasks;
 using System.Diagnostics;
 
-#if NETSTANDARD2_0
```
Comment (ppekrol) [inline src/Raven.Embedded/EmbeddedServer.cs]: revert

File: src/Raven.Embedded/EmbeddedServer.cs [hunk 4]
```diff
@@ -246,25 +242,11 @@ private async Task<(Uri ServerUrl, Process ServerProcess)> RunServer()
 
             process.Exited += (sender, e) => ServerProcessExited?.Invoke(sender, new ServerProcessExitedEventArgs());
 
-            bool domainBind;
 
-#if NETSTANDARD2_0
```
Comment (ppekrol) [inline src/Raven.Embedded/EmbeddedServer.cs]: revert

File: src/Raven.Embedded/EmbeddedServer.cs [hunk 5]
```diff
@@ -246,25 +242,11 @@ private async Task<(Uri ServerUrl, Process ServerProcess)> RunServer()
...
             AssemblyLoadContext.Default.Unloading += c =>
             {
                 ShutdownServerProcess(process);
             };
-            domainBind = true;
-#endif
-#if NET461
```
Comment (ppekrol) [inline src/Raven.Embedded/EmbeddedServer.cs]: revert

File: src/Raven.Embedded/Raven.Embedded.csproj [hunk 6]
```diff
@@ -4,6 +4,7 @@
     <Version></Version>
     <Description>RavenDB Embedded library to run ravendb in embedded way</Description>
     <Authors>Hibernating Rhinos</Authors>
+    <TargetFrameworks>net5.0;netcoreapp3.1</TargetFrameworks>
```
Comment (ppekrol) [inline src/Raven.Embedded/Raven.Embedded.csproj]: revert

File: src/Raven.Embedded/Raven.Embedded.csproj [hunk 7]
```diff
@@ -14,12 +15,6 @@
     <Configurations>Debug;Release;Validate</Configurations>
   </PropertyGroup>
 
-  <PropertyGroup Condition="'$(OS)' == 'Windows_NT'">
```
Comment (ppekrol) [inline src/Raven.Embedded/Raven.Embedded.csproj]: revert

File: src/Raven.Embedded/Raven.Embedded.csproj [hunk 8]
```diff
@@ -49,7 +44,4 @@
     </PackageReference>
   </ItemGroup>
 
-  <ItemGroup Condition="$(TargetFramework) == 'netstandard2.0'">
```
Comment (ppekrol) [inline src/Raven.Embedded/Raven.Embedded.csproj]: revert

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 9]
```diff
@@ -42,7 +43,7 @@ public class IncomingReplicationHandler : IDisposable
     {
         private readonly DocumentDatabase _database;
         private readonly TcpClient _tcpClient;
-        private readonly Stream _stream;
+        private Stream _stream;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: bring back 'readonly' ?

File: src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs [hunk 10]
```diff
@@ -1202,4 +1198,6 @@ public static class ReplicationMessageType
         public const string Heartbeat = "Heartbeat";
         public const string Documents = "Documents";
     }
+
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs]: formatting

File: src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs [hunk 11]
```diff
@@ -322,7 +322,7 @@ public bool ExecuteReplicationOnce(TcpConnectionOptions tcpConnectionOptions, Ou
                         _parent._lastSentDocumentEtag = _lastEtag;
                         _parent._lastDocumentSentTime = DateTime.UtcNow;
                         var changeVector = wasInterrupted ? null : DocumentsStorage.GetDatabaseChangeVector(documentsContext);
-                        _parent.SendHeartbeat(changeVector);
+                        _parent.SendHeartbeat(changeVector);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs]: formatting

File: src/Raven.TestDriver/Raven.TestDriver.csproj [hunk 12]
```diff
@@ -4,20 +4,14 @@
     <Description></Description>
     <Title>RavenDB TestDriver</Title>
     <Product>RavenDB TestDriver</Product>
+    <TargetFrameworks>net5.0;netcoreapp3.1</TargetFrameworks>
```
Comment (ppekrol) [inline src/Raven.TestDriver/Raven.TestDriver.csproj]: revert

File: src/Raven.TestDriver/Raven.TestDriver.csproj [hunk 13]
```diff
@@ -4,20 +4,14 @@
...
     <Authors>Hibernating Rhinos</Authors>
     <AssemblyName>Raven.TestDriver</AssemblyName>
     <PackageId>RavenDB.TestDriver</PackageId>
     <PackageTags>ravendb;client;database;nosql;doc db</PackageTags>
     <PackageLicenseFile>LICENSE.txt</PackageLicenseFile>
     <CodeAnalysisRuleSet>..\..\RavenDB.ruleset</CodeAnalysisRuleSet>
   </PropertyGroup>
-  
-  <PropertyGroup Condition="'$(OS)' == 'Windows_NT'">
```
Comment (ppekrol) [inline src/Raven.TestDriver/Raven.TestDriver.csproj]: revert

File: src/Sparrow/Sparrow.csproj [hunk 14]
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>net5.0;netcoreapp3.1;netcoreapp2.1;netstandard2.1;netstandard2.0</TargetFrameworks>
```
Comment (ppekrol) [inline src/Sparrow/Sparrow.csproj]: revert

File: src/Sparrow/Threading/DisposeOnce.cs [hunk 15]
```diff
@@ -212,6 +213,7 @@ public async Task DisposeAsync()
                     // is over. This call to Wait may throw with an
                     // AggregateException
                     await localState.Item2.Task.ConfigureAwait(false);
+                    return;
```
Comment (ppekrol) [inline src/Sparrow/Threading/DisposeOnce.cs]: I would need a comment why this change was made

File: src/Sparrow/Threading/DisposeOnce.cs [hunk 16]
```diff
@@ -179,6 +179,7 @@ public DisposeOnceAsync(Func<Task> action)
                 throw new NotSupportedException("Unknown operation mode: " + typeof(TOperationMode));
             }
             _operationModeData = default;
+            _operationModeData.Initialize();
```
Comment (ppekrol) [inline src/Sparrow/Threading/DisposeOnce.cs]: I would need a comment why this change was made

File: src/Sparrow/Utils/ConcatStream.cs [hunk 17]
```diff
@@ -4,9 +4,9 @@
 using System.Threading;
 using System.Threading.Tasks;
 
-namespace Raven.Client.Documents.Operations.Attachments
+namespace Sparrow.Utils
 {
-    internal class ConcatStream : Stream
+    public class ConcatStream : Stream
```
Comment (ppekrol) [inline src/Sparrow/Utils/ConcatStream.cs]: internal

File: src/Sparrow/Utils/DynamicNativeLibraryResolver.cs [hunk 18]
```diff
@@ -3,7 +3,7 @@
 using System.Reflection;
 using System.Runtime.InteropServices;
 
-namespace Sparrow.Server
+namespace Sparrow.Utils
 {
     public static class DynamicNativeLibraryResolver
```
Comment (ppekrol) [inline src/Sparrow/Utils/DynamicNativeLibraryResolver.cs]: internal... and why it was moved?

File: src/Sparrow/Utils/IncorrectDllException.cs [hunk 19]
```diff
@@ -1,6 +1,6 @@
 ﻿using System;
 
-namespace Sparrow.Server.Platform
+namespace Sparrow.Utils
 {
     public class IncorrectDllException : Exception
```
Comment (ppekrol) [inline src/Sparrow/Utils/IncorrectDllException.cs]: internal... and why it was moved?

File: src/Sparrow/Utils/ReadWriteCompressedStream.cs [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+using System.Buffers;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Sparrow.Json;
+using Sparrow.Threading;
+
+namespace Sparrow.Utils
+{
+    public class ReadWriteCompressedStream : Stream
```
Comment (ppekrol) [inline src/Sparrow/Utils/ReadWriteCompressedStream.cs]: internal

File: src/Sparrow/Utils/ReadWriteCompressedStream.cs [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+
+            await _inner.DisposeAsync().ConfigureAwait(false);
+
+            if (_input != null)
+                await _input.DisposeAsync().ConfigureAwait(false);
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            _dispose.DisposeAsync().GetAwaiter().GetResult();
```
Comment (ppekrol) [inline src/Sparrow/Utils/ReadWriteCompressedStream.cs]: Since this can be used in the Client... use AsyncHelpers.RunSync

File: src/Sparrow/Utils/ZstdLib.cs [hunk 22]
```diff
@@ -2,17 +2,17 @@
 using System.Diagnostics;
 using System.IO;
 using System.Runtime.InteropServices;
-using Sparrow.Server;
+using Sparrow.Platform;
 
-namespace Voron.Data.Tables
+namespace Sparrow.Utils
 {
     public static unsafe class ZstdLib
```
Comment (ppekrol) [inline src/Sparrow/Utils/ZstdLib.cs:10]: internal

File: src/Sparrow/Utils/ZstdStream.cs [hunk 23]
```diff
@@ -0,0 +1,361 @@
+﻿using System;
+using System.Buffers;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Sparrow.Threading;
+
+namespace Sparrow.Utils
+{
+    public class ZstdStream : Stream
```
Comment (ppekrol) [inline src/Sparrow/Utils/ZstdStream.cs]: internal

**Comment by ppekrol** [comment]
Please also rebase and edit commit messages to avoid "WIP" and "temp" names

File: src/Sparrow/Utils/ZstdStream.cs [hunk 24]
```diff
@@ -0,0 +1,361 @@
...
+            _compression = compression;
+            _disposeOnce = new DisposeOnceAsync<SingleAttempt>(DisposeInternal);
+        }
+
+        public static ZstdStream Compress(Stream stream) => new ZstdStream(stream, compression: true);
+        public static ZstdStream Decompress(Stream stream) => new ZstdStream(stream, compression: false);
+
+        public override bool CanRead => true;
+        public override bool CanSeek => false;
+        public override bool CanWrite => true;
```
Comment (ayende) [inline src/Sparrow/Utils/ZstdStream.cs]: `CanRead` and `CanWrite` should take their cue from `_compression`

File: src/Sparrow/Utils/ZstdStream.cs [hunk 25]
```diff
@@ -0,0 +1,361 @@
...
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            return Read(new Span<byte>(buffer, offset, count));
+        }
+
+        public override int Read(Span<byte> buffer)
+        {
+            if (_readSemaphore.Wait(10_000) == false)
```
Comment (ayende) [inline src/Sparrow/Utils/ZstdStream.cs]: Time should be configurable, and much smaller by default (1,500 ms)

File: src/Sparrow/Utils/ZstdStream.cs [hunk 26]
```diff
@@ -0,0 +1,361 @@
...
+                throw new InvalidOperationException("Couldn't acquire write lock for 10 seconds!");
```
Comment (ayende) [inline src/Sparrow/Utils/ZstdStream.cs]: Here you need to give a _detailed_ explanation. Right now this is going to be a badly understood error

Comment (ayende) [reply src/Sparrow/Utils/ZstdStream.cs]: Also, need to throw `TimeoutException`

File: src/Sparrow/Utils/ZstdStream.cs [hunk 27]
```diff
@@ -0,0 +1,361 @@
...
+                    _decompressionInput = new Memory<byte>(_tempBuffer, 0, _decompressionInput.Length + read);
+                    read = DecompressStep(buffer);
+                    if (read != 0)
+                        return read;
+                }
+            }
+            catch (ObjectDisposedException)
+            {
+                // object was already disposed, so nothing to do...
+            }
```
Comment (ayende) [inline src/Sparrow/Utils/ZstdStream.cs]: Why are you hiding the exception here?

File: src/Sparrow/Utils/ZstdStream.cs [hunk 28]
```diff
@@ -0,0 +1,361 @@
...
+            finally { _readSemaphore.Release(); }
+
+            return 0;
```
Comment (ayende) [inline src/Sparrow/Utils/ZstdStream.cs]: dead code

File: src/Sparrow/Utils/ZstdStream.cs (hunk 25 above)
Comment (ayende) [reply src/Sparrow/Utils/ZstdStream.cs]: That said, we only need to worry about this for the compression context itself, no?
In that case, let's move the scope of the semaphore to the compression / decompress step.
They are bounded and do not involve I/O, so that is much better.

File: src/Sparrow/Utils/ZstdStream.cs [hunk 29]
```diff
@@ -0,0 +1,361 @@
...
+        }
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            return ReadAsync(new Memory<byte>(buffer, offset, count), cancellationToken).AsTask();
+        }
+
+        public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
+        {
+            await _readSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
```
Comment (ayende) [inline src/Sparrow/Utils/ZstdStream.cs]: We'll be changing this, but note that this is NOT consistent with the `Read` method

File: src/Sparrow/Utils/ZstdStream.cs [hunk 30]
```diff
@@ -0,0 +1,361 @@
...
+                    _decompressionInput = new Memory<byte>(_tempBuffer, 0, _decompressionInput.Length + read);
+                    read = DecompressStep(buffer.Span);
+                    if (read != 0)
+                        return read;
+                }
+            }
+            catch (ObjectDisposedException)
+            {
+                // object was already disposed, so nothing to do...
+            }
```
Comment (ayende) [inline src/Sparrow/Utils/ZstdStream.cs]: Same

File: src/Sparrow/Utils/ZstdStream.cs [hunk 31]
```diff
@@ -0,0 +1,361 @@
...
+                {
+                    var (outputBytes, inputBytes, _) = CompressStep(buffer, ZstdLib.ZSTD_EndDirective.ZSTD_e_continue);
+                    buffer = buffer.Slice(inputBytes);
+                    _inner.Write(_tempBuffer, 0, outputBytes);
+                }
+            }
+            catch (ObjectDisposedException)
+            {
+                // object was already disposed, so nothing to do...
+            }
```
Comment (ayende) [inline src/Sparrow/Utils/ZstdStream.cs]: Same

File: src/Sparrow/Utils/ZstdStream.cs [hunk 32]
```diff
@@ -0,0 +1,361 @@
...
+                {
+                    var (outputBytes, inputBytes, _) = CompressStep(buffer.Span, ZstdLib.ZSTD_EndDirective.ZSTD_e_continue);
+                    buffer = buffer.Slice(inputBytes);
+                    await _inner.WriteAsync(_tempBuffer, 0, outputBytes, cancellationToken).ConfigureAwait(false);
+                }
+            }
+            catch (ObjectDisposedException)
+            {
+                // object was already disposed, so nothing to do...
+            }
```
Comment (ayende) [inline src/Sparrow/Utils/ZstdStream.cs]: Same

File: src/Sparrow/Utils/ZstdStream.cs [hunk 33]
```diff
@@ -0,0 +1,361 @@
...
+                        break;
+                    _inner.Write(_tempBuffer, 0, outputBytes);
+                }
+
+                _inner.Flush();
+            }
+            catch (ObjectDisposedException)
+            {
+                // object was already disposed, so nothing to do...
+            }
```
Comment (ayende) [inline src/Sparrow/Utils/ZstdStream.cs]: same

File: src/Sparrow/Utils/ZstdStream.cs [hunk 34]
```diff
@@ -0,0 +1,361 @@
...
+        {
+            try
+            {
+                if (_compression == false)
+                {
+                    await _readSemaphore.WaitAsync().ConfigureAwait(false);
+                }
+                else
+                {
+                    await _writeSemaphore.WaitAsync().ConfigureAwait(false);
```
Comment (ayende) [inline src/Sparrow/Utils/ZstdStream.cs]: I don't understand the logic here

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Raven.Client.csproj (hunk 2 above)
Comment (shiranshalom) [reply src/Raven.Client/Raven.Client.csproj]: Done

File: src/Raven.Embedded/EmbeddedServer.cs (hunk 3 above)
Comment (shiranshalom) [reply src/Raven.Embedded/EmbeddedServer.cs]: Done

File: src/Raven.Embedded/EmbeddedServer.cs (hunk 4 above)
Comment (shiranshalom) [reply src/Raven.Embedded/EmbeddedServer.cs]: Done

File: src/Raven.Embedded/EmbeddedServer.cs (hunk 5 above)
Comment (shiranshalom) [reply src/Raven.Embedded/EmbeddedServer.cs]: Done

File: src/Sparrow/Utils/ReadWriteCompressedStream.cs (hunk 20 above)
Comment (shiranshalom) [reply src/Sparrow/Utils/ReadWriteCompressedStream.cs]: Done

File: src/Sparrow/Utils/ZstdStream.cs (hunk 23 above)
Comment (shiranshalom) [reply src/Sparrow/Utils/ZstdStream.cs]: Done

File: src/Sparrow/Utils/ReadWriteCompressedStream.cs [hunk 35]
```diff
@@ -0,0 +1,191 @@
...
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state)
+        {
```
Comment (ppekrol) [inline src/Sparrow/Utils/ReadWriteCompressedStream.cs:106]: Please check the warnings

File: src/Raven.Client/Documents/Subscriptions/SubscriptionWorker.cs [hunk 36]
```diff
@@ -226,14 +226,22 @@ await requestExecutor.ExecuteAsync(_redirectNode, null, context, command, should
 
                 var databaseName = _store.GetDatabase(_dbName);
 
+                bool compressionSupport = false;
+#if NETCOREAPP3_1_OR_GREATER
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Subscriptions/SubscriptionWorker.cs]: Where this is defined? Probably best place would be Directory.Build.props
Also it is better to do it the other way around since the number of unsupported target frameworks is more or less constant and number of supported ones will grow.

File: src/Sparrow/Utils/ReadWriteCompressedStream.cs (hunk 35 above)
Comment (shiranshalom) [reply src/Sparrow/Utils/ReadWriteCompressedStream.cs:106]: Done

File: src/Raven.Client/Documents/Subscriptions/SubscriptionWorker.cs (hunk 36 above)
Comment (shiranshalom) [reply src/Raven.Client/Documents/Subscriptions/SubscriptionWorker.cs]: Done

File: src/Sparrow/Utils/ZstdStream.cs (hunk 26 above)
Comment (shiranshalom) [reply src/Sparrow/Utils/ZstdStream.cs]: Done

File: src/Sparrow/Utils/ZstdStream.cs (hunk 25 above)
Comment (shiranshalom) [reply src/Sparrow/Utils/ZstdStream.cs]: Done

File: src/Sparrow/Utils/ZstdStream.cs (hunk 24 above)
Comment (shiranshalom) [reply src/Sparrow/Utils/ZstdStream.cs]: Done
